includes/multi-currency/MultiCurrency.php

```diff
@@ -1184,27 +1154,106 @@ private function register_admin_scripts() {
 		);
 	}
 
+	/**
+	 * Validates the given currency code.
+	 *
+	 * @param   string $currency_code  The currency code to check validity.
+	 *
+	 * @return  string|false  Returns back the currency code in uppercase letters if it's valid, or `false` if not.
+	 */
+	public function validate_currency_code( $currency_code ) {
+		return in_array( strtoupper( $currency_code ), array_keys( $this->available_currencies ), true )
+		? strtoupper( $currency_code )
+		: false;
+	}
+
+	/**
+	 * Get simulation params from querystring and activate when needed
+	 *
+	 * @return  void
+	 */
+	public function possible_simulation_activation() {
+		// This is required in the MC onboarding simulation iframe.
+		$this->simulation_params = $this->get_multi_currency_onboarding_simulation_variables();
+		if ( ! $this->is_simulation_enabled() ) {
+			return;
+		}
+		// Modify the page links to deliver required params in the simulation.
+		$this->add_simulation_params_to_preview_urls();
+		$this->simulate_client_currency();
+	}
+
+	/**
+	 * Enables simulation of client browser currency.
+	 *
+	 * @return  void
+	 */
+	public function simulate_client_currency() {
+		if ( ! $this->simulation_params['enable_auto_currency'] || ! $this->simulation_params['client_currency'] ) {
+			return;
+		}
+		$simulation_currency = $this->simulation_params['client_currency'];
+
+		// Simulate client currency from geolocation.
+		add_filter(
+			'wcpay_multi_currency_get_selected_currency',
+			function( $selected_currency ) use ( $simulation_currency ) {
```

Comment:
I am not sure that I understand how that works, the `$selected_currency` variable is not used here. 
Would you explain it a bit to me, please?

Reply:
I'm not sure which part you didn't understand, but sure, I'll explain both the possibilities. 

### 1. You didn't understand how the filter works

The filter provides the current value (`$selected_currency`) as the parameter for us to tinker on it and return the modified version back, thats how `add_filter` works, but we don't use that and return our `$simulation_currency` instead. 

I'll explain with an example.

```php 

$p = 2;

$p = apply_filters(`p_value_modifiers`, $p);

echo $p;
````

If there's no filters hooked to the `p_value_modifiers` hook, it'll echo 2. If we add this:

```php
add_filter(`p_value_modifiers`, function($old_p){ return $p * 2; });
```

it'll echo 4, because the $p will be modified with that filter.

But here, we're not modifying the data, we're replacing it, so think it like this:

```php
add_filter(`p_value_modifiers`, function($old_p){ return 5; });
```

it'll echo 5.

### 2. You didn't understand why we keep the `$selected_currency` as a parameter and gave the variable as an outside variable;

The filters are probably applied with `call_user_func`, and that'll expect the callback to have the matching parameter count with the filter itself. Otherwise it'll throw an error.

Reply:
Thanks for a detailed explanation, I really appreciate it! 

I meant 

> 2. You didn't understand why we keep the $selected_currency as a parameter and gave the variable as an outside variable;

Yes, they are applied with `call_user_func `, I've re-read the code once again, now it's clear 👍